[Skip to content](https://github.com/tryigit/CleveresTricky/actions/runs/22466888783/job/65074836648?pr=350#start-of-content)
[CleveresTricky](https://github.com/tryigit/CleveresTricky)
Repository navigation
[Code](https://github.com/tryigit/CleveresTricky)
[Issues](https://github.com/tryigit/CleveresTricky/issues)
[Actions](https://github.com/tryigit/CleveresTricky/actions)
[Back to pull request #350](https://github.com/tryigit/CleveresTricky/pull/350)
Implement Race Condition Engine and Stealth Features #822
Annotations
1 error
build
failed now in 3m 25s
2s
4s
0s
8s
8s
2m 50s
9s
  ld.lld: error: undefined reference due to --no-allow-shlib-undefined: android::RefBase::onFirstRef()
  >>> referenced by /home/runner/work/CleveresTricky/CleveresTricky/module/build/intermediates/cxx/RelWithDebInfo/604e1u6u/obj/arm64-v8a/libbinder.so
  
  ld.lld: error: undefined reference due to --no-allow-shlib-undefined: android::RefBase::onLastStrongRef(void const*)
  >>> referenced by /home/runner/work/CleveresTricky/CleveresTricky/module/build/intermediates/cxx/RelWithDebInfo/604e1u6u/obj/arm64-v8a/libbinder.so
  
  ld.lld: error: undefined reference due to --no-allow-shlib-undefined: android::RefBase::onIncStrongAttempted(void const*)
  >>> referenced by /home/runner/work/CleveresTricky/CleveresTricky/module/build/intermediates/cxx/RelWithDebInfo/604e1u6u/obj/arm64-v8a/libbinder.so
  clang++: error: linker command failed with exit code 1 (use -v to see invocation)
  [12/18] Building CXX object CMakeFiles/cleverestricky.dir/binder_interceptor.cpp.o
  [13/18] Building CXX object external/CMakeFiles/lsplt.dir/LSPlt/lsplt/src/main/jni/elf_util.cc.o
  [14/18] Building CXX object external/CMakeFiles/lsplt.dir/LSPlt/lsplt/src/main/jni/lsplt.cc.o
  ninja: build stopped: subcommand failed.
  
  C++ build system [build] failed while executing:
      /usr/local/lib/android/sdk/cmake/3.31.5/bin/ninja \
        -C \
        /home/runner/work/CleveresTricky/CleveresTricky/module/.cxx/RelWithDebInfo/604e1u6u/arm64-v8a \
        binder \
        cleverestricky \
        inject \
        keystore_race_daemon \
        utils
    from /home/runner/work/CleveresTricky/CleveresTricky/module

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org./

BUILD FAILED in 8s
19 actionable tasks: 19 executed
Error: Process completed with exit code 1.
0s
0s
0s
0s
0s
0s
0s
0s
0s
0s
1s
0s
0s
Footer
© 2026 GitHub, Inc.
Footer navigation
[Terms](https://docs.github.com/site-policy/github-terms/github-terms-of-service)
[Privacy](https://docs.github.com/site-policy/privacy-policies/github-privacy-statement)
[Security](https://github.com/security)
[Status](https://www.githubstatus.com/)
[Community](https://github.community/)
[Docs](https://docs.github.com/)
[Contact](https://support.github.com/?tags=dotcom-footer)
Manage cookies
Do not share my personal information

--- a/module/src/main/cpp/binder/stub_utils.cpp
+++ b/module/src/main/cpp/binder/stub_utils.cpp
@@ -2,61 +2,65 @@
 #include "utils/RefBase.h"
 #include "utils/String16.h"
 
+// Ensure these symbols are exported from libutils.so so libbinder.so can use them.
+// Without this, -fvisibility=hidden causes undefined references in shared libraries.
+#define EXPORT [[gnu::visibility("default")]]
+
 namespace android {
-    void RefBase::incStrong(const void *id) const {
+    EXPORT void RefBase::incStrong(const void *id) const {
 
     }
 
-    void RefBase::incStrongRequireStrong(const void *id) const {
+    EXPORT void RefBase::incStrongRequireStrong(const void *id) const {
 
     }
 
-    void RefBase::decStrong(const void *id) const {
+    EXPORT void RefBase::decStrong(const void *id) const {
 
     }
 
-    void RefBase::forceIncStrong(const void *id) const {
+    EXPORT void RefBase::forceIncStrong(const void *id) const {
 
     }
 
-    RefBase::weakref_type* RefBase::createWeak(const void* id) const {
+    EXPORT RefBase::weakref_type* RefBase::createWeak(const void* id) const {
         return nullptr;
     }
 
-    RefBase::weakref_type* RefBase::getWeakRefs() const {
+    EXPORT RefBase::weakref_type* RefBase::getWeakRefs() const {
         return nullptr;
     }
 
-    RefBase::RefBase(): mRefs(nullptr) {}
-    RefBase::~RefBase() {}
+    EXPORT RefBase::RefBase(): mRefs(nullptr) {}
+    EXPORT RefBase::~RefBase() {}
 
-    void RefBase::onFirstRef() {}
-    void RefBase::onLastStrongRef(const void* id) {}
-    bool RefBase::onIncStrongAttempted(const void* id) { return false; }
+    EXPORT void RefBase::onFirstRef() {}
+    EXPORT void RefBase::onLastStrongRef(const void* id) {}
+    EXPORT bool RefBase::onIncStrongAttempted(const void* id) { return false; }
 
-    RefBase* RefBase::weakref_type::refBase() const { return nullptr; }
+    EXPORT RefBase* RefBase::weakref_type::refBase() const { return nullptr; }
 
-    void RefBase::weakref_type::incWeak(const void* id) {}
-    void RefBase::weakref_type::incWeakRequireWeak(const void* id) {}
-    void RefBase::weakref_type::decWeak(const void* id) {}
+    EXPORT void RefBase::weakref_type::incWeak(const void* id) {}
+    EXPORT void RefBase::weakref_type::incWeakRequireWeak(const void* id) {}
+    EXPORT void RefBase::weakref_type::decWeak(const void* id) {}
 
     // acquires a strong reference if there is already one.
-    bool RefBase::weakref_type::attemptIncStrong(const void* id) { return false; }
+    EXPORT bool RefBase::weakref_type::attemptIncStrong(const void* id) { return false; }
 
     // acquires a weak reference if there is already one.
     // This is not always safe. see ProcessState.cpp and BpBinder.cpp
     // for proper use.
-    bool RefBase::weakref_type::attemptIncWeak(const void* id) { return false; }
+    EXPORT bool RefBase::weakref_type::attemptIncWeak(const void* id) { return false; }
 
-    void sp_report_race() {}
+    EXPORT void sp_report_race() {}
 
-    String16::String16() {}
+    EXPORT String16::String16() {}
 
-    String16::String16(const String16 &o) {}
+    EXPORT String16::String16(const String16 &o) {}
 
-    String16::String16(String16 &&o) noexcept {}
+    EXPORT String16::String16(String16 &&o) noexcept {}
 
-    String16::String16(const char *o) {}
+    EXPORT String16::String16(const char *o) {}
 
-    String16::~String16() {}
+    EXPORT String16::~String16() {}
 }

